scene/main/canvas_item.cpp

```diff
@@ -60,7 +60,7 @@ Transform2D CanvasItem::_edit_get_transform() const {
 
 bool CanvasItem::is_visible_in_tree() const {
 	ERR_READ_THREAD_GUARD_V(false);
-	return visible && parent_visible_in_tree;
+	return visible && parent_visible_in_tree && is_node_active_in_tree();
```

Comment:
`is_node_active_in_tree()` seems to be checking activity recursively in its ancestors. It should be a cached value, like `parent_visible_in_tree`.

Reply:
I can do that, though the way that node_3d.cpp handles visibility in tree is actually to traverse each parent, seen [here](https://github.com/godotengine/godot/blob/26d1577f3985363faab48a65e9a0d9eed0e26d86/scene/3d/node_3d.cpp#L825). Is there a reason its like that? And this one was modeled after the `process_owner` system i.e. it doesn't check all parents one by one but just the topmost active parent in the chain thanks to this code:
```
if (data.parent && data.node_active) {
	data.active_owner = data.parent->data.active_owner;
} 
```
I could instead turn this into the way that canvas_item.cpp handles visibility with a bool, would that be preferred?

Reply:
CanvasItem visibility was optimized to avoid tree lookup, 3D code was not updated yet (assuming it's possible; there might be differences preventing it).

The `process_owner` doesn't make recursive calls, so the method does not get more expensive based on node depth. But this is possible, because the owner determines only process mode; the processing itself is determined based on mode. It can't be applied to activity, because it depends on whether the parent is active.

For optimal solution, `active` should match the CanvasItem visibility implementation.

Reply:
Done, thanks! Let me know if this implementation looks better 😄